xbox/matrix.h

```diff
@@ -0,0 +1,318 @@
+            m[12] * m[3] * m[6];
+
+  inv[10] = m[0]  * m[5] * m[15] - 
+            m[0]  * m[7] * m[13] - 
+            m[4]  * m[1] * m[15] + 
+            m[4]  * m[3] * m[13] + 
+            m[12] * m[1] * m[7] - 
+            m[12] * m[3] * m[5];
+
+  inv[14] = -m[0]  * m[5] * m[14] + 
+             m[0]  * m[6] * m[13] + 
+             m[4]  * m[1] * m[14] - 
+             m[4]  * m[2] * m[13] - 
+             m[12] * m[1] * m[6] + 
+             m[12] * m[2] * m[5];
+
+  inv[3] = -m[1] * m[6] * m[11] + 
+            m[1] * m[7] * m[10] + 
+            m[5] * m[2] * m[11] - 
+            m[5] * m[3] * m[10] - 
+            m[9] * m[2] * m[7] + 
+            m[9] * m[3] * m[6];
+
+  inv[7] = m[0] * m[6] * m[11] - 
+           m[0] * m[7] * m[10] - 
+           m[4] * m[2] * m[11] + 
+           m[4] * m[3] * m[10] + 
+           m[8] * m[2] * m[7] - 
+           m[8] * m[3] * m[6];
+
+  inv[11] = -m[0] * m[5] * m[11] + 
+             m[0] * m[7] * m[9] + 
+             m[4] * m[1] * m[11] - 
+             m[4] * m[3] * m[9] - 
+             m[8] * m[1] * m[7] + 
+             m[8] * m[3] * m[5];
+
+  inv[15] = m[0] * m[5] * m[10] - 
+            m[0] * m[6] * m[9] - 
+            m[4] * m[1] * m[10] + 
+            m[4] * m[2] * m[9] + 
+            m[8] * m[1] * m[6] - 
+            m[8] * m[2] * m[5];
+
+  det = m[0] * inv[0] + m[1] * inv[4] + m[2] * inv[8] + m[3] * inv[12];
+
+  if (det == 0)
+      return false;
+
+  det = 1.0 / det;
+
+  for (i = 0; i < 16; i++)
+      invOut[i] = inv[i] * det;
+
+  return true;
+}
+
+#define A(row,col)  a[(col<<2)+row]
+#define B(row,col)  b[(col<<2)+row]
+#define P(row,col) product[(col<<2)+row]
+static void matmul4( GLfloat *product, const GLfloat *a, const GLfloat *b )
+{
+  GLint i;
+  for (i = 0; i < 4; i++) {
+    const GLfloat ai0=A(i,0),  ai1=A(i,1),  ai2=A(i,2),  ai3=A(i,3);
+    P(i,0) = ai0 * B(0,0) + ai1 * B(1,0) + ai2 * B(2,0) + ai3 * B(3,0);
+    P(i,1) = ai0 * B(0,1) + ai1 * B(1,1) + ai2 * B(2,1) + ai3 * B(3,1);
+    P(i,2) = ai0 * B(0,2) + ai1 * B(1,2) + ai2 * B(2,2) + ai3 * B(3,2);
+    P(i,3) = ai0 * B(0,3) + ai1 * B(1,3) + ai2 * B(2,3) + ai3 * B(3,3);
+  }
+}
+#undef A
+#undef B
+#undef P
+
+void ortho(float* r, GLdouble left, GLdouble right, GLdouble bottom, GLdouble top, GLdouble nearval, GLdouble farval) {
+  //FIXME: Multiply this onto the existing stack - don't overwrite
+
+  GLfloat m[16];
+
+#define M(row,col)  m[col*4+row]
+  M(0,0) = 2.0F / (right-left);
+  M(0,1) = 0.0F;
+  M(0,2) = 0.0F;
+  M(0,3) = -(right+left) / (right-left);
+
+  M(1,0) = 0.0F;
+  M(1,1) = 2.0F / (top-bottom);
+  M(1,2) = 0.0F;
+  M(1,3) = -(top+bottom) / (top-bottom);
+
+  M(2,0) = 0.0F;
+  M(2,1) = 0.0F;
+  M(2,2) = -2.0F / (farval-nearval);
+  M(2,3) = -(farval+nearval) / (farval-nearval);
+
+  M(3,0) = 0.0F;
+  M(3,1) = 0.0F;
+  M(3,2) = 0.0F;
+  M(3,3) = 1.0F;
+#undef M
+
+  memcpy(r, m, sizeof(m));
+}
+
+void _math_matrix_rotate( GLfloat *m, GLfloat angle, GLfloat x, GLfloat y, GLfloat z )
```

Comment:
Review this - I took out some optimization to keep the code short. I also moved the initialization into the caller as done for other functions already (WHY?!).